{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Copyright **`(c)`** 2023 Giovanni Squillero `<giovanni.squillero@polito.it>`  \n",
    "[`https://github.com/squillero/computational-intelligence`](https://github.com/squillero/computational-intelligence)  \n",
    "Free for personal or classroom use; see [`LICENSE.md`](https://github.com/squillero/computational-intelligence/blob/master/LICENSE.md) for details.  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# LAB9\n",
    "\n",
    "Write a local-search algorithm (eg. an EA) able to solve the *Problem* instances 1, 2, 5, and 10 on a 1000-loci genomes, using a minimum number of fitness calls. That's all.\n",
    "\n",
    "### Deadlines:\n",
    "\n",
    "* Submission: Sunday, December 3 ([CET](https://www.timeanddate.com/time/zones/cet))\n",
    "* Reviews: Sunday, December 10 ([CET](https://www.timeanddate.com/time/zones/cet))\n",
    "\n",
    "Notes:\n",
    "\n",
    "* Reviews will be assigned  on Monday, December 4\n",
    "* You need to commit in order to be selected as a reviewer (ie. better to commit an empty work than not to commit)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [],
   "source": [
    "from random import choices, random, randint\n",
    "import numpy as np\n",
    "from copy import deepcopy\n",
    "\n",
    "import lab9_lib\n",
    "\n",
    "# GOAL: maximize fitness, minimize calls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0010101110000110100001111100101110010101001010110100001011011101111011100100100100101100101001001010: 10.89%\n",
      "1000110011110110010100011011010000111110101010011001100001110110101010101101001110100100000100111001: 11.89%\n",
      "0100111001010010011101111011101000100001011011110000101010010110001100110011100011001100101011000101: 24.91%\n",
      "0110010100110101011111100010001101111100001000010010001100100000100100001010100001100010101000011010: 29.35%\n",
      "1100000111001100001011001100101011010100110101101101101111111100000000100110111111100101000011101011: 22.89%\n",
      "1000111110011101101010111000101000101010011100100101111011011110101000000111000110110000010010111001: 10.79%\n",
      "0111001001110010110001001100010001001000001010100111010110111101100001111001101010101101001010000110: 10.89%\n",
      "0111011010111110001101010011101110101110000101011011000110110111000011001110111001101100010011011111: 14.57%\n",
      "1101010000010101101100010100101100101111100010010001011001101000100101101111111110110100100101110110: 12.69%\n",
      "1011101111111101001010100000000001100010100011000110101010101100100111110111101011001001011011101110: 35.02%\n",
      "0110000001110111111111010100110100111000111011000101111101100000011000101011110111110001000011010010: 11.79%\n",
      "1000100110011001011001001101101001000110111100100110101010110110101110000000111000000010101001110001: 9.90%\n",
      "0000011010001101000101011101001010011010110110110000100001101111111000001010101000001101111011010010: 20.91%\n",
      "1110011001010100111101101111100100010010110001001111101110001110111110001101101001101101011000110100: 26.80%\n",
      "0001110001011001100010110000101101101001011011100110010001110000000111111110011001110000001111101110: 10.89%\n",
      "1011010111010110100001101101001001011100100110000011100110101010110111100100001101101001100110001100: 10.79%\n",
      "1100110001101101000010000001111101010001011010101111011100100000000011100110110000101101101101111001: 11.69%\n",
      "0010011011010111001000001011010110100010111010011010110010000001110110100010101011100101010000111011: 9.90%\n",
      "1010000011110100110101110100001101011100111101111011101011010100101010100100001000000101001001100111: 31.93%\n",
      "1111110111101101010110111100010110101101001011010101110110011101011011111101010010101011101011101100: 14.46%\n",
      "20\n"
     ]
    }
   ],
   "source": [
    "fitness = lab9_lib.make_problem(5)\n",
    "for n in range(20):\n",
    "    ind = choices([0, 1], k=100)\n",
    "    print(f\"{''.join(str(g) for g in ind)}: {fitness(ind):.2%}\")\n",
    "\n",
    "print(fitness.calls)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Our code below"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "metadata": {},
   "outputs": [],
   "source": [
    "POPULATION_SIZE = 100\n",
    "OFFSPRING_SIZE = 100\n",
    "LOCI = 1000\n",
    "BIT_FLIP_PROBABILITY = 0.15\n",
    "NUM_GENERATION = 100\n",
    "\n",
    "\n",
    "class Individual:\n",
    "    def __init__(self):\n",
    "        self.genotype = choices([0, 1], k=LOCI)\n",
    "        self.fitness = float(\"-inf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mutation / recombination or both\n",
    "def parent_selection(\n",
    "    population: list[Individual], tournament_size: int\n",
    ") -> list[Individual]:\n",
    "    # we also want to take the last best one.\n",
    "    parents_idx = np.random.choice(\n",
    "        range(len(population)), size=tournament_size, replace=False\n",
    "    )\n",
    "    parents = [population[idx] for idx in parents_idx]\n",
    "    return sorted(parents, key=lambda i: i.fitness, reverse=True)[:2]\n",
    "\n",
    "\n",
    "def one_cut_xover(ind1: Individual, ind2: Individual) -> Individual:\n",
    "    cut_point = randint(0, LOCI - 1)\n",
    "    new_ind = Individual()\n",
    "    new_ind.genotype = ind1.genotype[:cut_point] + ind2.genotype[cut_point:]\n",
    "    assert len(new_ind.genotype) == LOCI\n",
    "    return new_ind\n",
    "\n",
    "\n",
    "def two_cut_xover(ind1: Individual, ind2: Individual) -> Individual:\n",
    "    first_cut_point = randint(0, LOCI / 2)\n",
    "    second_cut_point = randint(LOCI / 2, (LOCI - 1))\n",
    "    new_ind = Individual()\n",
    "    new_ind.genotype = (\n",
    "        ind1.genotype[:first_cut_point]\n",
    "        + ind2.genotype[first_cut_point:second_cut_point]\n",
    "        + ind1.genotype[second_cut_point:]\n",
    "    )\n",
    "    assert len(new_ind.genotype) == LOCI\n",
    "    return new_ind\n",
    "\n",
    "\n",
    "# This xover function returns a child whose genome\n",
    "# is created proportionally to the fitenss of parents\n",
    "def uniform_cut_xover(ind1: Individual, ind2: Individual) -> Individual:\n",
    "    # p1 = ind1.fitness / (ind1.fitness + ind2.fitness)\n",
    "    p1 = 0.5\n",
    "    gene = [\n",
    "        np.random.choice([ind1.genotype[i], ind2.genotype[i]], p=[p1, 1 - p1])\n",
    "        for i in range(LOCI)\n",
    "    ]\n",
    "    new_ind = Individual()\n",
    "    new_ind.genotype = gene\n",
    "    return new_ind\n",
    "\n",
    "\n",
    "def mutate(parent: Individual) -> Individual:\n",
    "    new_offspring = deepcopy(parent)\n",
    "    for i in range(LOCI):\n",
    "        if random() < BIT_FLIP_PROBABILITY:\n",
    "            new_offspring.genotype[i] = int(not new_offspring.genotype[i])\n",
    "    return new_offspring\n",
    "\n",
    "\n",
    "def offspring_generation(\n",
    "    parent1: Individual, parent2: Individual, mutation_probability: int\n",
    ") -> Individual:\n",
    "    if random() < mutation_probability:\n",
    "        # mutation\n",
    "        return mutate(parent1)\n",
    "    else:\n",
    "        # cross_over\n",
    "        return uniform_cut_xover(parent1, parent2)\n",
    "\n",
    "\n",
    "def ea() -> Individual:\n",
    "    # starting pouplation of POPULATION_SIZE individuals\n",
    "    population = [Individual() for _ in range(POPULATION_SIZE)]\n",
    "    for p in population:\n",
    "        p.fitness = fitness(p.genotype)\n",
    "    old_best_fitness = 0\n",
    "    best_fitness = population[0].fitness\n",
    "    # for _ in range(NUM_GENERATION)\n",
    "    gen = 0\n",
    "    mutation_probability = 0.5\n",
    "    tournament_size = 10\n",
    "    while best_fitness > old_best_fitness or gen < NUM_GENERATION:\n",
    "        old_best_fitness = best_fitness\n",
    "        num_of_better_offspring = 0\n",
    "        # best_old_one = best_one\n",
    "        best_parents = parent_selection(population, tournament_size)\n",
    "        for _ in range(OFFSPRING_SIZE):\n",
    "            offspring = offspring_generation(\n",
    "                best_parents[0], best_parents[1], mutation_probability\n",
    "            )\n",
    "            offspring.fitness = fitness(offspring.genotype)\n",
    "            population.extend([offspring])\n",
    "            # tracking the number of offspring better than the best old one.\n",
    "            if offspring.fitness > best_parents[0].fitness:\n",
    "                num_of_better_offspring += 1\n",
    "\n",
    "        population.sort(key=lambda i: i.fitness, reverse=True)\n",
    "        # always keep the first POPULATION_SIZE best individuals\n",
    "        population = population[:POPULATION_SIZE]\n",
    "\n",
    "        # Self-adapting the values.\n",
    "        # TODO favour mutation insted of recombination if the fitness it's not high (<0.65).\n",
    "        if num_of_better_offspring >= 0.2 * OFFSPRING_SIZE:\n",
    "            # If we are able to generate 20% of offsprings better than the father,\n",
    "            # We should explore more.\n",
    "            tournament_size += 1\n",
    "            mutation_probability *= 1.2\n",
    "        else:\n",
    "            # We are probably around a good point, we should favour exploitation.\n",
    "            tournament_size = tournament_size - 1 if tournament_size > 5 else 5\n",
    "            mutation_probability /= 1.2\n",
    "\n",
    "        best_fitness = population[0].fitness\n",
    "        gen += 1\n",
    "\n",
    "        print(f\"gen #{gen}, fitness: {best_fitness}\")\n",
    "\n",
    "    return population[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gen #1, fitness: 0.53\n",
      "gen #2, fitness: 0.53\n",
      "gen #3, fitness: 0.53\n",
      "gen #4, fitness: 0.53\n",
      "gen #5, fitness: 0.534\n",
      "gen #6, fitness: 0.534\n",
      "gen #7, fitness: 0.538\n",
      "gen #8, fitness: 0.538\n",
      "gen #9, fitness: 0.538\n",
      "gen #10, fitness: 0.538\n",
      "gen #11, fitness: 0.538\n",
      "gen #12, fitness: 0.538\n",
      "gen #13, fitness: 0.54\n",
      "gen #14, fitness: 0.54\n",
      "gen #15, fitness: 0.54\n",
      "gen #16, fitness: 0.544\n",
      "gen #17, fitness: 0.544\n",
      "gen #18, fitness: 0.544\n",
      "gen #19, fitness: 0.544\n",
      "gen #20, fitness: 0.556\n",
      "gen #21, fitness: 0.556\n",
      "gen #22, fitness: 0.556\n",
      "gen #23, fitness: 0.556\n",
      "gen #24, fitness: 0.556\n",
      "gen #25, fitness: 0.556\n",
      "gen #26, fitness: 0.556\n",
      "gen #27, fitness: 0.556\n",
      "gen #28, fitness: 0.556\n",
      "gen #29, fitness: 0.556\n",
      "gen #30, fitness: 0.556\n",
      "gen #31, fitness: 0.556\n",
      "gen #32, fitness: 0.556\n",
      "gen #33, fitness: 0.556\n",
      "gen #34, fitness: 0.556\n",
      "gen #35, fitness: 0.556\n",
      "gen #36, fitness: 0.556\n",
      "gen #37, fitness: 0.558\n",
      "gen #38, fitness: 0.558\n",
      "gen #39, fitness: 0.558\n",
      "gen #40, fitness: 0.558\n",
      "gen #41, fitness: 0.562\n",
      "gen #42, fitness: 0.562\n",
      "gen #43, fitness: 0.568\n",
      "gen #44, fitness: 0.568\n",
      "gen #45, fitness: 0.568\n",
      "gen #46, fitness: 0.568\n",
      "gen #47, fitness: 0.568\n",
      "gen #48, fitness: 0.576\n",
      "gen #49, fitness: 0.576\n",
      "gen #50, fitness: 0.576\n",
      "gen #51, fitness: 0.576\n",
      "gen #52, fitness: 0.576\n",
      "gen #53, fitness: 0.576\n",
      "gen #54, fitness: 0.576\n",
      "gen #55, fitness: 0.576\n",
      "gen #56, fitness: 0.576\n",
      "gen #57, fitness: 0.576\n",
      "gen #58, fitness: 0.576\n",
      "gen #59, fitness: 0.576\n",
      "gen #60, fitness: 0.576\n",
      "gen #61, fitness: 0.576\n",
      "gen #62, fitness: 0.576\n",
      "gen #63, fitness: 0.576\n",
      "gen #64, fitness: 0.58\n",
      "gen #65, fitness: 0.58\n",
      "gen #66, fitness: 0.58\n",
      "gen #67, fitness: 0.58\n",
      "gen #68, fitness: 0.582\n",
      "gen #69, fitness: 0.582\n",
      "gen #70, fitness: 0.582\n",
      "gen #71, fitness: 0.582\n",
      "gen #72, fitness: 0.586\n",
      "gen #73, fitness: 0.586\n",
      "gen #74, fitness: 0.586\n",
      "gen #75, fitness: 0.588\n",
      "gen #76, fitness: 0.592\n",
      "gen #77, fitness: 0.592\n",
      "gen #78, fitness: 0.592\n",
      "gen #79, fitness: 0.592\n",
      "gen #80, fitness: 0.592\n",
      "gen #81, fitness: 0.592\n",
      "gen #82, fitness: 0.592\n",
      "gen #83, fitness: 0.592\n",
      "gen #84, fitness: 0.592\n",
      "gen #85, fitness: 0.592\n",
      "gen #86, fitness: 0.592\n",
      "gen #87, fitness: 0.592\n",
      "gen #88, fitness: 0.592\n",
      "gen #89, fitness: 0.592\n",
      "gen #90, fitness: 0.592\n",
      "gen #91, fitness: 0.592\n",
      "gen #92, fitness: 0.592\n",
      "gen #93, fitness: 0.596\n",
      "gen #94, fitness: 0.596\n",
      "gen #95, fitness: 0.596\n",
      "gen #96, fitness: 0.598\n",
      "gen #97, fitness: 0.598\n",
      "gen #98, fitness: 0.6\n",
      "gen #99, fitness: 0.6\n",
      "gen #100, fitness: 0.6\n",
      "Best individual fitness: 0.6\n",
      "Fitness calls: 10100\n"
     ]
    }
   ],
   "source": [
    "# instance = [1, 2, 5, 10]\n",
    "instance = [2]\n",
    "for k in instance:\n",
    "    fitness = lab9_lib.make_problem(k)\n",
    "    print(f\"Best individual fitness: {ea().fitness}\")\n",
    "    print(f\"Fitness calls: {fitness.calls}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ci-fLJ3OwGs-py3.12",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
